File: server/datasets/python_source-code/Multiple-B1/organic_5.py
# Binary Tree Traversal - organic submission 5
# Iterative post-order using the two-stacks technique.

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


def insert(root, data):
    if root is None:
        return Node(data)
    if data < root.data:
        root.left = insert(root.left, data)
    else:
        root.right = insert(root.right, data)
    return root


def postorder_two_stacks(root):
    if root is None:
        return []
    stack1 = [root]
    stack2 = []
    while stack1:
        node = stack1.pop()
        stack2.append(node.data)
        if node.left:
            stack1.append(node.left)
        if node.right:
            stack1.append(node.right)
    stack2.reverse()
    return stack2


if __name__ == "__main__":
    root = None
    for d in [50, 30, 70, 20, 40, 60, 80]:
        root = insert(root, d)
    print("Post-order (two stacks):", postorder_two_stacks(root))

File: server/datasets/controlled_benchmarks/java/multi_student_cohorts/cohort_b1/organic_5.java
// Binary Tree Traversal - organic submission 5
// Iterative post-order using the two-stacks technique.

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Stack;

class Node {
    int data;
    Node left;
    Node right;

    public Node(int data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

class Solution {
    public static Node insert(Node root, int data) {
        if (root == null) {
            return new Node(data);
        }
        if (data < root.data) {
            root.left = insert(root.left, data);
        } else {
            root.right = insert(root.right, data);
        }
        return root;
    }

    public static List<Integer> postorderTwoStacks(Node root) {
        if (root == null) {
            return new ArrayList<>();
        }
        Stack<Node> stack1 = new Stack<>();
        List<Integer> stack2 = new ArrayList<>();
        stack1.push(root);
        while (!stack1.isEmpty()) {
            Node node = stack1.pop();
            stack2.add(node.data);
            if (node.left != null) {
                stack1.push(node.left);
            }
            if (node.right != null) {
                stack1.push(node.right);
            }
        }
        Collections.reverse(stack2);
        return stack2;
    }

    public static void main(String[] args) {
        Node root = null;
        for (int d : new int[]{50, 30, 70, 20, 40, 60, 80}) {
            root = insert(root, d);
        }
        System.out.println("Post-order (two stacks): " + postorderTwoStacks(root));
    }
}
